From 36030ed505b8a03cadc3669f37c848d8824db5e2 Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Thu, 13 Apr 2023 12:40:45 +0300
Subject: [PATCH 3/8] net: phy: dp83869: implement did_interrupt

Implement did_interrupt handler to determine which phy signaled
interrupt when using a shared interrupt signal.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 drivers/net/phy/dp83869.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/net/phy/dp83869.c b/drivers/net/phy/dp83869.c
index dcc1c5c0c121..2baa869124ba 100644
--- a/drivers/net/phy/dp83869.c
+++ b/drivers/net/phy/dp83869.c
@@ -194,6 +194,22 @@ static int dp83869_ack_interrupt(struct phy_device *phydev)
 	return 0;
 }
 
+static int dp83869_did_interrupt(struct phy_device *phydev)
+{
+	int isr, fisr;
+
+	// read ISR
+	isr = phy_read(phydev, MII_DP83869_ISR);
+
+	// read fiber ISR
+	fisr = phy_read(phydev, DP83860_FX_INT_STS);
+
+	if (isr || fisr & DP83860_FX_INT_STS_MASK)
+		return 1;
+
+	return 0;
+}
+
 static int dp83869_config_intr(struct phy_device *phydev)
 {
 	int micr_status = 0;
@@ -868,6 +884,7 @@ static struct phy_driver dp83869_driver[] = {
 		.ack_interrupt	= dp83869_ack_interrupt,
 		.config_intr	= dp83869_config_intr,
 		.read_status	= dp83869_read_status,
+		.did_interrupt	= dp83869_did_interrupt,
 
 		.get_tunable	= dp83869_get_tunable,
 		.set_tunable	= dp83869_set_tunable,
-- 
2.35.3

